sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala

```diff
@@ -563,15 +570,43 @@ class Analyzer(
       }
     }
 
+    private object EmptyGroupingIDExtractor {
+
+      private def hasEmptyGroupingID(aggExprs: Seq[Expression]): Boolean = {
+        aggExprs.exists { p => p.collectFirst { case GroupingID(Nil) => true }.isDefined }
+      }
+
+      def unapply(p: LogicalPlan): Option[Seq[Expression]] = p match {
+        case Aggregate(Seq(Cube(groupByExprs)), aggExprs, _) if hasEmptyGroupingID(aggExprs) =>
+          Some(constructGroupBy(cubeExprs(groupByExprs)))
+        case Aggregate(Seq(Rollup(groupByExprs)), aggExprs, _) if hasEmptyGroupingID(aggExprs) =>
+          Some(constructGroupBy(rollupExprs(groupByExprs)))
+        case x: GroupingSets if hasEmptyGroupingID(x.aggregations) =>
+          Some(constructGroupBy(x.selectedGroupByExprs))
+        case Filter(cond, child) if hasEmptyGroupingID(cond :: Nil) =>
+          Some(findGroupingExprs(child))
+        case Sort(order, _, child) if hasEmptyGroupingID(order) =>
+          Some(findGroupingExprs(child))
+        case _ =>
+          None
+      }
+    }
+
     // This require transformUp to replace grouping()/grouping_id() in resolved Filter/Sort
     def apply(plan: LogicalPlan): LogicalPlan = plan resolveOperatorsUp {
       case a if !a.childrenResolved => a // be sure all of the children are resolved.
 
+      // If a plan has grouping IDs with an empty group-by, we resolve them first
+      case p @ EmptyGroupingIDExtractor(groupByExprs) =>
+        p.transformExpressions {
+          case GroupingID(Nil) => GroupingID(groupByExprs)
+        }
+
```

Comment:
These change are refactoring or for different issue?

Reply:
I added this new rule for not refactoring but supporting string grouping IDs. This pr makes a `GroupingID` dataType depend on the number of group-by expressions (int for a few attributes and string for many ones). So, in case of the empty group-by case, we need to resolve it (by filling in group-by exprs) here before analyzing grouping analytics (cube, roll-up, and grouping set).